### DISCUSSION TITLE
Why Expo-av does not supports DRM Encrypted Video

### DISCUSSION THREAD

**Discussion by narendramanu** [expo/expo#19807] [Archived: Questions]
I am trying to play a video which is DRM encrypted using Widevine but Expo av is not able to render video and audio. Can this be achieved in any alternative ways using expo-av

**Comment by tslater** [comment]


**Reply by fanguitocoder** [reply]
Because they don't care?

**Comment by tslater** [comment]
Would the maintainers accept a PR that adds DRM support to expo-av on both iOS and Android? We're working directly with the Hollywood studios and have hard requirements for DRM, so we're working on it.

@sjchmiela @lukmccall  @EvanBacon  @terribleben

**Reply by JustJoostNL** [reply]
Any ETA? Would really love to see this!

**Reply by tslater** [reply]
ETA is 3 weeks.

**Comment by garfieldnate** [comment]
Yes, having DRM support would help anyone (including me) working with media services that require it as a matter of course, which I don't think is terribly uncommon.

**Comment by tsapeta** [comment]
Hey! Yeah @tslater, we would accept such a PR, but depending on how much complexity it brings to the entire package.
On the other hand, we're going to split `expo-av` into `expo-video` (and later `expo-audio` too) where I think we'll definitely need to support DRMs, but I can't give any estimation right now, so probably better to just add it to `expo-av` first.

**Comment by tslater** [comment]
Thanks @tsapeta. I've scoped this out and am hiring DRM experienced contractors to add DRM. We expect to submit PR within 3 weeks. I've also sent them the contribution guidelines.

**Reply by tsapeta** [reply]
Nice, thanks! Let me know if you/they need any help. I've already made a proof of concept of the new `expo-video` package on iOS, so maybe we can add this feature only there? That might be a little bit easier as it's written in Swift and we won't have to migrate this from `expo-av` in the future when it gets deprecated.

**Reply by CodeXpertsDev** [reply]
@tsapeta  Please share the POC for expo-video package on iOS, thanks.

**Reply by tslater** [reply]
@tsapeta Can you give us a url to the branch or area of code where the POC `expo-video` package is? Do you think it will be released in the next major expo version?

**Reply by JustJoostNL** [reply]
@tslater @CodeXpertsDev  https://github.com/expo/expo/tree/main/packages/expo-video

**Comment by mojavad** [comment]
@tslater @tsapeta Was wondering if there's any update on DRM support? It's something we could use on our project.

**Reply by tslater** [reply]
Working on it

**Reply by dgocoder** [reply]
Any update on this?

**Reply by JustJoostNL** [reply]
Still waiting for this as well.

**Reply by rgab1508** [reply]
@tslater any ETA?

**Reply by tslater** [reply]
My contractor says that they got Android/Widevine working. I haven't had time to test it quite yet. Can any of you try it out and me know if it works for you? He's starting on iOS/FairPlay next week
https://github.com/onizam95/expo-av-drm/tree/feat/add-drm,

**Reply by tslater** [reply]
Here's instructions on how to use it:
```
{
    uri: 'https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey/Manifest_1080p.mpd',
    drm: {
        type: 'widevine', //or DRMType.WIDEVINE
        licenseServer: 'https://drm-widevine-licensing.axtest.net/AcquireLicense',
        headers: {
            'X-AxDRM-Message':
'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ2ZXJzaW9uIjoxLCJjb21fa2V5X2lkIjoiYjMzNjRlYjUtNTFmN
i00YWUzLThjOTgtMzNjZWQ1ZTMxYzc4IiwibWVzc2FnZSI6eyJ0eXBlIjoiZW50aXRsZW1lbnRfbWVzc2FnZSIsImZp
cnN0X3BsYXlfZXhwaXJhdGlvbiI6NjAsInBsYXlyZWFkeSI6eyJyZWFsX3RpbWVfZXhwaXJhdGlvbiI6dHJ1ZX0sImt
leXMiOlt7ImlkIjoiOWViNDA1MGQtZTQ0Yi00ODAyLTkzMmUtMjdkNzUwODNlMjY2IiwiZW5jcnlwdGVkX2tleSI6Im
xLM09qSExZVzI0Y3Iya3RSNzRmbnc9PSJ9XX19.FAbIiPxX8BHi9RwfzD7Yn-wugU19ghrkBFKsaCPrZmU'
}, }
}
<VideoPlayer
        sources={[
            {uri:
'https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey/Manifest_1080p.mpd', headers:
{test: 'test11',drm : {type: 'widevine', licenseServer:
'https://drm-widevine-licensing.axtest.net/AcquireLicense', drmHeaders: {
              'X-AxDRM-Message':
'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ2ZXJzaW9uIjoxLCJjb21fa2V5X2lkIjoiYjMzNjRlYjUtNTFmN
i00YWUzLThjOTgtMzNjZWQ1ZTMxYzc4IiwibWVzc2FnZSI6eyJ0eXBlIjoiZW50aXRsZW1lbnRfbWVzc2FnZSIsImZp
cnN0X3BsYXlfZXhwaXJhdGlvbiI6NjAsInBsYXlyZWFkeSI6eyJyZWFsX3RpbWVfZXhwaXJhdGlvbiI6dHJ1ZX0sImt
leXMiOlt7ImlkIjoiOWViNDA1MGQtZTQ0Yi00ODAyLTkzMmUtMjdkNzUwODNlMjY2IiwiZW5jcnlwdGVkX2tleSI6Im
xLM09qSExZVzI0Y3Iya3RSNzRmbnc9PSJ9XX19.FAbIiPxX8BHi9RwfzD7Yn-wugU19ghrkBFKsaCPrZmU'
}}}}
]} />
```

**Reply by rgab1508** [reply]
Great! I can try this but I have no idea how to install this https://github.com/onizam95/expo-av-drm/tree/feat/add-drm

**Reply by tslater** [reply]
TBH me neither or I would have tried it. I'm assuming it's whatever the workflow of an expo contributor would be.

**Reply by tslater** [reply]
Something like this:
https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-download-and-setup

**Reply by onizam95** [reply]
Hello @rgab1508 
here at @CodeXpertsDev  we developed in to keep the changes as minimum as possible. Please follow the expo contribution link which @tslater  mentioned above from my forked branch. 
run the BareExpo app and in code In `native-components-list/src/AV/VideoScreen.tsx` add the drm stream as @tslater mentioned above. 
to Pass the drmConfigs We have used the  headers map or source Prop. 
this is just for android right now and needs some refinement and testing. Once we are done with iOS as well we will create a proper PR with complete instructions. 
let me know if you need any help in setup

**Reply by onizam95** [reply]
you can also see the changes in git history for a better reference

**Reply by rgab1508** [reply]
cool, will give it a try it @tslater @onizam95

**Reply by onizam95** [reply]
Hey @tslater @rgab1508 
We have added the support for ios too in the latest commit. 
You will have to pass your configs like this for ios
`<VideoPlayer
        sources={[
            {uri: '', headers: {drm : {type: 'fairplay', licenseServer: '', certificateUrl: '', base64Cert: false }}}
        ]}
      />`

**Comment by tslater** [comment]
Please let us know how it goes!

**Comment by tslater** [comment]
I finally got around to testing this and it's working on Android and iOS. The team that helped get it going is now working on a PR. @rgab1508 @mojavad @narendramanu

**Reply by JustJoostNL** [reply]
Great! Can't wait! 😄

**Reply by dgocoder** [reply]
Looking forward to trying this out

**Comment by JustJoostNL** [comment]
@onizam95 Is there the possibility to use a custom getLicense request just like in react-native-video?

**Reply by onizam95** [reply]
currently no because as in another thread of discussion Expo contributors mentioned that they wanted minimum changes. I started with the inspiration from react-native-video but then switched my approach and now the license setup is native(under the hood).
But yes this is possible.

**Reply by JustJoostNL** [reply]
Ah okay. Will this still be added (in the future), or isn’t this planned?

**Reply by onizam95** [reply]
I am working on creating the PR draft. I will discuss it with the reviewer. IF they are okay with it then we can see what we can do about it. but not planned at the moment.

**Reply by JustJoostNL** [reply]
That’s alright, thanks.

**Reply by onizam95** [reply]
https://github.com/expo/expo/pull/25868
here is the PR link just FYI.

**Reply by dgocoder** [reply]
in the interim is there an easy way to use this while waiting for PR to get merged?

**Reply by onizam95** [reply]
@dgocoder unfortunately we tried but we were unable to find such way. I believe expo is not designed to use anythign other than released sdk or canary release. I am not sure if I missed something but i tried a lot for it.
@tsapeta can you comment on this please and also on the PR.

**Reply by tsapeta** [reply]
I'll take a look at the PR soon, but as far as I can see right now it's a bit outdated as we've recently moved some code to different location.

Until then, I think the best way to let others test these changes is to publish your version of the package to npm (under your organization/user scope). I believe that those changes are compatible with SDK 49 and 50 (development builds and bare workflow) as it doesn't change anything internal to Expo.

**Reply by onizam95** [reply]
@tsapeta  I will update the PR,
about the npm I considered the option but @tslater wants to use it in Expo (expo-av ) prod builds, there is no way to do that?
because otherwise react-native-video is there with DRM.

**Reply by tsapeta** [reply]
@onizam95 The only place where it can't be used is Expo Go, but it's no longer recommended for building production-ready apps. The way we recommend is to use development builds for development and then using EAS build or `npx expo prebuild` command to generate the native code and then submit to the stores. You can use any native module in development builds.

**Reply by tslater** [reply]
@tsapeta , what I'm looking for (and I think everyone else is too) is a way to publish a module that can be used in an expo project with a custom dev client (not expo go). Is there a relatively easy way to do that before this gets merged?

**Reply by tsapeta** [reply]
@tslater Yes, just change the `name` in the package.json of your forked package and publish it to npm under your user or organization scope, like `@tslater/expo-av`. I guess that should work, but if it doesn't, I'll need to fix that as it might be annoying also for other contributors.

**Reply by onizam95** [reply]
@tsapeta  this is not working as when i copy expo-av from packages and create it as a new package it gives error on resolving expo core modules. 
I discussed it with a few devs in discord channel as well and they got same issue.

**Reply by tslater** [reply]
@tsapeta Can you show the command you ran and the error output?

**Reply by onizam95** [reply]
Hi @tsapeta 
 I have updated my [PR](https://github.com/expo/expo/pull/25868), Please have a look now,
 Also I tried to just copy expo-av folder and change its name in package.json and publish to npm. but while running it gives this error
 `Error: Could not find module 'ExpoModulesCore'`

**Comment by tslater** [comment]
Posting the PR link here:
https://github.com/expo/expo/pull/25868

**Reply by onizam95** [reply]
NEW/UPDATED PR [LINK](https://github.com/expo/expo/pull/26433)
@tsapeta

**Comment by tslater** [comment]
Hey everyone: @fanguitocoder @narendramanu @mojavad @rgab1508 @JustJoostNL @dgocoder

I published the package and have tested on iOS. I need to test Android. Please help us test!
Should be the same process as installing `expo-av` except instead you need to install `@tsl8r/expo-av`
Here's the NPM link:
https://www.npmjs.com/package/@tsl8r/expo-av

Warning: The typescript types seem off...but I'll post an example.

For iOS:
```
<VIdeo 
  source={
  uri: 'https://yourdomain.com/manifest.m3u8',
  headers: {
    drm:
    {
      type: 'fairplay',
      licenseServer: 'https://your-url.com,
      certificateUrl: 'https://you-certificate-domain.com/somecertname.der',
      base64Cert: false
    }
  }
}}
```

**Reply by JustJoostNL** [reply]
@onizam95 Since the PR got closed and [the other PR](https://github.com/expo/expo/pull/26465) is for expo-video, which won't be available before SDK 51. Would you be able to fix the types of your package so we can use it easier with TS?

**Reply by JustJoostNL** [reply]
@onizam95 @tslater 

I just tested the implementation from above, unfortunately, I am getting the following error:
`Video error The AVPlayerItem instance has failed with the error code -1002 and domain "NSURLErrorDomain".`

I am using this code (ignore weird indentation):

```tsx
<Video
  shouldPlay
  useNativeControls
  source={{
    uri: "https://d1v1gqjxxxxu1.cloudfront.net/media/xxxxxxxx/xxxxx/xxxx.m3u8",
    headers: {
      //@ts-ignore
      drm: {
         type: "fairplay",
         licenseServer: "https://license.vdocipher.com/auth/",
         certificateUrl: "https://themes.vdocipher.com/xxxxxxxxx/xxxxxx/fairplay.cer",
         base64Cert: false,
      },
     },
  }}
   style={{
     width: 300,
      height: 200,
  }}
     onError={(error) => {
       console.log("Video error", error);
      }}
/>
```

Any idea what's causing the error here?

**Comment by tslater** [comment]
@onizam95 So I'm able to build the package, import it into my project etc. It runs in iOS, but in android I get a runtime error (below).

My assumption here is that it is a version mismatch between the current state of the the https://github.com/onizam95/expo-av-drm/tree/feat/add-drm fork and  expo-modules-core. Assuming that's correct, how do I figure out what to use?

```
01-12 16:30:46.048 26098 26098 I ExpoDevMenu: Delegate's context was loaded.
01-12 16:30:46.049 26098 26098 E SensorManager: registerListenerImpl sensorName:bmi160 Accelerometer Non-wakeup,isWakeUpSensor:false,callingApp: com.wordscenes.dev,callingPid:26098,callingUid:10238
01-12 16:30:46.068 26098 26337 I ExpoModulesCore: ✅ AppContext was initialized
01-12 16:30:46.075 26098 26337 W unknown:ReactContext: initializeMessageQueueThreads() is called.
01-12 16:30:47.931 26098 26329 I ExpoModulesCore: ✅ JSI interop was installed
01-12 16:30:47.938 26098 26329 I dev.expo.updates: {"timestamp":1705095047937,"message":"UpdatesModule: getConstants called","code":"None","level":"info"}
01-12 16:30:47.990 26098 26329 I ExpoModulesCore: ✅ Constants were exported
01-12 16:30:48.801 26098 26329 V SoLoader: libreanimated.so not found on /data/app/~~ZnMLjI1QNuVfa6UrbWx9uw==/com.wordscenes.dev-CGTL4mt9wpyHvWkGyK9PEA==/lib/arm64
01-12 16:30:48.816 26098 26329 D SoLoader: libreanimated.so found on /data/app/~~ZnMLjI1QNuVfa6UrbWx9uw==/com.wordscenes.dev-CGTL4mt9wpyHvWkGyK9PEA==/base.apk!/lib/arm64-v8a
01-12 16:30:49.716 26098 26329 I ReactNativeJS: Running "main" with {"rootTag":31}
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577] JNI DETECTED ERROR IN APPLICATION: JNI GetObjectRefType called with pending exception expo.modules.kotlin.exception.MissingTypeConverter: Cannot find type converter for 'expo.modules.core.arguments.ReadableArguments'.
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.types.TypeConverter expo.modules.kotlin.types.TypeConverterProviderImpl.obtainTypeConverter(kotlin.reflect.KType) (TypeConverterProvider.kt:142)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.types.TypeConverter expo.modules.kotlin.types.AnyType$converter$2.invoke() (AnyType.kt:59)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at java.lang.Object expo.modules.kotlin.types.AnyType$converter$2.invoke() (AnyType.kt:58)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at java.lang.Object kotlin.SynchronizedLazyImpl.getValue() (LazyJVM.kt:74)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.types.TypeConverter expo.modules.kotlin.types.AnyType.getConverter() (AnyType.kt:58)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.jni.ExpectedType expo.modules.kotlin.types.AnyType.getCppRequiredTypes() (AnyType.kt:64)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void expo.modules.kotlin.functions.AsyncFunction.attachToJSObject(expo.modules.kotlin.AppContext, expo.modules.kotlin.jni.JavaScriptModuleObject) (AsyncFunction.kt:64)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.jni.JavaScriptModuleObject expo.modules.kotlin.jni.JavaScriptModuleObject.initUsingObjectDefinition(expo.modules.kotlin.AppContext, expo.modules.kotlin.objects.ObjectDefinitionData) (JavaScriptModuleObject.kt:41)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.jni.JavaScriptModuleObject expo.modules.kotlin.ModuleHolder$jsObject$2.invoke() (ModuleHolder.kt:32)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at java.lang.Object expo.modules.kotlin.ModuleHolder$jsObject$2.invoke() (ModuleHolder.kt:26)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at java.lang.Object kotlin.SynchronizedLazyImpl.getValue() (LazyJVM.kt:74)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.jni.JavaScriptModuleObject expo.modules.kotlin.ModuleHolder.getJsObject() (ModuleHolder.kt:26)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at expo.modules.kotlin.jni.JavaScriptModuleObject expo.modules.kotlin.jni.JSIInteropModuleRegistry.getJavaScriptModuleObject(java.lang.String) (JSIInteropModuleRegistry.kt:82)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void com.facebook.jni.NativeRunnable.run() (NativeRunnable.java:-2)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void android.os.Handler.handleCallback(android.os.Message) (Handler.java:938)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:99)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(android.os.Message) (MessageQueueThreadHandler.java:27)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void android.os.Looper.loop() (Looper.java:233)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run() (MessageQueueThreadImpl.java:228)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]   at void java.lang.Thread.run() (Thread.java:923)
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577] 
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]     in call to GetObjectRefType
01-12 16:30:50.867 26098 26329 F .wordscenes.de: java_vm_ext.cc:577]     from void com.facebook.jni.NativeRunnable.run()
```

UPDATE: I tried forcing  "expo-modules-core": "1.7.0" using resolutions because of this in the changelog:
[Android] Add type converter for the ReadableArguments class to allow backward compatibility with older modules. (https://github.com/expo/expo/pull/24137 by [@lukmccall](https://github.com/lukmccall))

But now the bundle won't build and load:
`TypeError: (0 , _paths).getEntryPoint is not a function`

**Comment by tslater** [comment]
Ok, I was able to get past all build errors up grading to expo sdk 50. I'm getting runtime errors with the standalone package though:

```
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: ❌ Cannot set the 'ExpoVideoView' prop on the 'expo.modules.av.video.VideoViewWrapper{3a8c357 V.E...... ......I. 0,0-0,0 #cb}'
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: expo.modules.kotlin.exception.PropSetException: Cannot set prop 'source' on view 'class expo.modules.av.video.VideoViewWrapper'
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: → Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Map.containsKey(java.lang.Object)' on a null object reference
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.ConcreteViewProp.set(ConcreteViewProp.kt:19)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.ViewManagerWrapperDelegate.updateProperties(ViewManagerWrapperDelegate.kt:69)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.GroupViewManagerWrapper.updateProperties(GroupViewManagerWrapper.kt:23)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.GroupViewManagerWrapper.updateProperties(GroupViewManagerWrapper.kt:12)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.ViewManager.createViewInstance(ViewManager.java:181)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.ViewManager.createView(ViewManager.java:108)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.NativeViewHierarchyManager.createView(NativeViewHierarchyManager.java:281)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.UIViewOperationQueue$CreateViewOperation.execute(UIViewOperationQueue.java:194)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.dispatchPendingNonBatchedOperations(UIViewOperationQueue.java:1121)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:1092)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1350)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.view.Choreographer.doCallbacks(Choreographer.java:1149)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.view.Choreographer.doFrame(Choreographer.java:1040)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1333)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.os.Handler.handleCallback(Handler.java:938)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.os.Handler.dispatchMessage(Handler.java:99)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.os.Looper.loop(Looper.java:233)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at android.app.ActivityThread.main(ActivityThread.java:8063)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at java.lang.reflect.Method.invoke(Native Method)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: Caused by: expo.modules.kotlin.exception.UnexpectedException: java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Map.containsKey(java.lang.Object)' on a null object reference
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.ConcreteViewProp.set(ConcreteViewProp.kt:35)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	... 23 more
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Map.containsKey(java.lang.Object)' on a null object reference
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.player.SimpleExoPlayerData.<init>(SimpleExoPlayerData.java:78)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.player.PlayerData.createUnloadedPlayerData(PlayerData.java:222)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.video.VideoView.setSource(VideoView.java:363)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.video.VideoView.setSource(VideoView.java:332)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.video.VideoViewModule$definition$1$1$4.invoke(VideoViewModule.kt:44)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.av.video.VideoViewModule$definition$1$1$4.invoke(VideoViewModule.kt:43)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	at expo.modules.kotlin.views.ConcreteViewProp.set(ConcreteViewProp.kt:21)
01-13 11:44:49.507  7340  7340 E ExpoModulesCore: 	... 23 more

```

@onizam95  have you seen this before?

**Reply by onizam95** [reply]
@tslater I was getting the same errors and discussed in the discord group but unfortunately no luck with these. 
I am not sure how we can use expo-av as standalone package with ejecting from expo.  or even with npm expo-av as custom package.

**Reply by tslater** [reply]
I was able to get past this by using SDK 50 beta for my base project I am installing into

**Reply by tslater** [reply]
Try this:
```
npx create-expo-app --template blank@beta
yarn add [@tsl8r/expo-av](https://www.npmjs.com/package/@tsl8r/expo-av)
```

Then I run `eas build --profile=development --local --platform ios`
which works.

Then I run `eas build --profile=development --local --platform android` is not working for me

**Reply by onizam95** [reply]
There might be new changes in SDK 50. I was trying with SDK 49. I will try with sdk 50 in the morning and will get back to you

**Reply by tslater** [reply]
How did sdk 50 go? @onizam95

**Reply by onizam95** [reply]
Hi @tslater 

Just Now I have fixed the android source crash. 

it was because when no request headers were passed in source for normal video sources. Please take latest pull. will get back to you soon as well for standalone package progress with sdk 50

**Comment by onizam95** [comment]
NEW/UPDATED PR [LINK](https://github.com/expo/expo/pull/26433)
